subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -0,0 +1,2031 @@
+{
+	int err;
+
+	*cme_err = CME_ERROR_AG_FAILURE;
+
+	if (!is_char(buf, '\r')) {
+		return -EINVAL;
+	}
+
+	if (ag->selected_codec_id == 0) {
+		return -EINVAL;
+	}
+
+	if (!(ag->hf_codec_ids & BIT(ag->selected_codec_id))) {
+		return -EINVAL;
+	}
+
+	if (ag->call_state == BT_HFP_CALL_TERMINATE) {
+		return -EINVAL;
+	}
+
+	if (ag->sco_chan.sco != NULL) {
+		return -EINVAL;
+	}
+
+	err = hfp_ag_next_step(ag, bt_hfp_ag_audio_connection, NULL);
+
+	return 0;
+}
+
+static void bt_hfp_ag_unit_codec_conn_setup(struct bt_hfp_ag *ag, void *user_data)
+{
+	int err = hfp_ag_open_sco(ag);
+
+	if (err) {
+		bt_hfp_ag_call_reject(ag, user_data);
+	}
+}
+
+static int bt_hfp_ag_bcs_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *cme_err)
+{
+	int err;
+	uint32_t number;
+
+	*cme_err = CME_ERROR_AG_FAILURE;
+
+	if (!is_char(buf, '=')) {
+		return -EINVAL;
+	}
+
+	err = get_number(buf, &number);
+	if (err != 0) {
+		return err;
+	}
+
+	if (!is_char(buf, '\r')) {
+		return -EINVAL;
+	}
+
+	if (!atomic_test_bit(ag->flags, BT_HFP_AG_CODEC_CONN)) {
+		return -EINVAL;
+	}
+
+	if ((ag->selected_codec_id != number) ||
+	    (!(ag->hf_codec_ids & BIT(ag->selected_codec_id)))) {
+		err = -EINVAL;
+	}
+
+	if (err == 0) {
+		atomic_clear_bit(ag->flags, BT_HFP_AG_CODEC_CONN);
+		atomic_clear_bit(ag->flags, BT_HFP_AG_CODEC_CHANGED);
+		err = hfp_ag_next_step(ag, bt_hfp_ag_unit_codec_conn_setup, NULL);
+	} else {
+		if (ag->call_state != BT_HFP_CALL_TERMINATE) {
+			(void)hfp_ag_next_step(ag, bt_hfp_ag_unit_call_terminate, NULL);
+		}
+	}
+
+	return err;
+}
+
+static void bt_hfp_ag_unit_call_outgoing(struct bt_hfp_ag *ag, void *user_data)
+{
+	int err = bt_hfp_ag_outgoing(ag, ag->number);
+
+	if (err != 0) {
+		LOG_ERR("Fail to send err :(%d)", err);
+	}
+}
+
+static int bt_hfp_ag_atd_handler(struct bt_hfp_ag *ag, struct net_buf *buf, enum at_cme *cme_err)
+{
+	int err;
+	char *number = NULL;
+	bool is_memory_dial = false;
+
+	*cme_err = CME_ERROR_AG_FAILURE;
+
+	if (buf->data[buf->len - 1] != '\r') {
+		*cme_err = CME_ERROR_OPERATION_NOT_ALLOWED;
+		return -EINVAL;
+	}
+
+	if (is_char(buf, '>')) {
+		is_memory_dial = true;
+	}
+
+	if ((buf->len - 1) > CONFIG_BT_HFP_AG_PHONE_NUMBER_MAX_LEN) {
+		*cme_err = CME_ERROR_DIAL_STRING_TO_LONG;
+		return -EINVAL;
+	}
+
+	buf->data[buf->len - 1] = '\0';
+
+	if (is_memory_dial) {
+		if (bt_ag && bt_ag->memory_dial) {
+			err = bt_ag->memory_dial(ag, &buf->data[0], &number);
+			if ((err != 0) || (number == NULL)) {
+				*cme_err = CME_ERROR_INVALID_INDEX;
+				return -EINVAL;
+			}
+		}
+	} else {
+		number = &buf->data[0];
+	}
+
+	if (strlen(number) == 0) {
+		return -EINVAL;
+	}
+
+	if (strlen(number) > CONFIG_BT_HFP_AG_PHONE_NUMBER_MAX_LEN) {
+		*cme_err = CME_ERROR_DIAL_STRING_TO_LONG;
+		return -EINVAL;
+	}
+
+	if (ag->call_state != BT_HFP_CALL_TERMINATE) {
+		return -EBUSY;
+	}
+
+	memcpy(ag->number, number, strlen(number));
+	ag->number[strlen(number)] = '\0';
```

Comment:
I guess `strncpy(ag->number, number, sizeof(ag->number))` might be more appropriate here? Calling `strlen()` multiple times on the same variable isn't particularly efficient either (right now you're doing it four times).

Reply:
strncpy will write all memory of ag->number. It is no point in writing additional memory space.
I updated it to 
`len = strlen(number)`
`memcpy(ag->number, number, len + 1)`